pkg/bubbletea/bubbletea_pagination_util.go

```diff
@@ -0,0 +1,174 @@
+package bubbletea
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/flyteorg/flytectl/pkg/filters"
+	"github.com/flyteorg/flytectl/pkg/printer"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+	"github.com/kataras/tablewriter"
+	"github.com/landoop/tableprinter"
+	"github.com/yalp/jsonpath"
+)
+
+type DataCallback func(filter filters.Filters) []proto.Message
+
+type PrintableProto struct{ proto.Message }
+
+const (
+	defaultMsgPerBatch = 100
+	defaultMsgPerPage  = 10
+	pagePerBatch       = defaultMsgPerBatch / defaultMsgPerPage
+)
+
+var (
+	firstBatchIndex int32 = 1
+	lastBatchIndex  int32 = 10
+	batchLen              = make(map[int32]int)
+
+	// Callback function used to fetch data from the module that called bubbletea pagination.
+	callback DataCallback
+	// The header of the table
+	listHeader []printer.Column
+
+	marshaller = jsonpb.Marshaler{
+		Indent: "\t",
+	}
+)
+
+func (p PrintableProto) MarshalJSON() ([]byte, error) {
+	buf := new(bytes.Buffer)
+	err := marshaller.Marshal(buf, p.Message)
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+func extractRow(data interface{}, columns []printer.Column) []string {
+	if columns == nil || data == nil {
+		return nil
+	}
+
+	tableData := make([]string, 0, len(columns))
+	for _, c := range columns {
+		out, err := jsonpath.Read(data, c.JSONPath)
+		if err != nil || out == nil {
+			out = ""
+		}
+		s := fmt.Sprintf("%s", out)
+		if c.TruncateTo != nil {
+			t := *c.TruncateTo
+			if len(s) > t {
+				s = s[:t]
+			}
+		}
+		tableData = append(tableData, s)
+	}
+	return tableData
+}
+
+func projectColumns(rows []interface{}, column []printer.Column) [][]string {
+	responses := make([][]string, 0, len(rows))
+	for _, row := range rows {
+		responses = append(responses, extractRow(row, column))
+	}
+	return responses
+}
+
+func printTable(m *pageModel, start int, end int) (string, error) {
+	curShowMessage := m.items[start:end]
+	printableMessages := make([]*PrintableProto, 0, len(curShowMessage))
+	for _, m := range curShowMessage {
+		printableMessages = append(printableMessages, &PrintableProto{Message: m})
+	}
+
+	jsonRows, err := json.Marshal(printableMessages)
+	if err != nil {
+		return "", fmt.Errorf("failed to marshal proto messages")
+	}
+
+	var rawRows []interface{}
```

Comment:
Is this reusable?
https://github.com/flyteorg/flytectl/blob/64853531cb35f2a6214ff1cc82e9d39fc6109c08/pkg/printer/printer.go#L115